README.md
# spektrum

### Dennis Evangelista
This library provides mbed support for Spektrum serial receivers. So far, the library has been tested on a [Spektrum SPM9745 DSMX remote receiver](https://www.horizonhobby.com/SPM9745?KPID=SPM9745&CAWELAID=320011980001425764&CAGPSPN=pla&CAAGID=37619207031&CATCI=pla-518519028442&gclid=CjwKCAiAz7TfBRAKEiwAz8fKOFgsklymSCgwfoY6IgzJ0decXHqctabFSpUVYezkrSFwuWUkJqiS4RoCbcwQAvD_BwE) (since v1.0.0), and will probably be tested on a SPM9645 in the near future. 

## Obtaining the library
Source code is revision controlled using Git and hosted in a Github repository.
```bash
git clone ssh://git@github.com/devangel77b/Spektrum.git
git pull
```
The ```dev``` branch is bleeding edge; ```master``` branch is reasonably up to date. 

## Using the library
The main part of the library are two classes for ```Spektrum``` (receiver) and ```BindPlug```, which allows binding a transmitter to the receiver. In addition, a convenience macro for converting 0-2048 values to PwmOut pulsewidths (600-2100us, centered at 900us) is provided (```SPEKTRUM_COUNT2US()```). 

## Note on servo position (0-2048) and pwm pulse width
The specification for Spektrum Remote Receiver Interfacing, rev A, 12 Apr 2016, has a note about servo position ranges:

> A full range of servo position data ranges from 0-1024 or 0-2045 depending on the bind type. These limits are equivalent to a +/-150% travel setting in AirWare. This full range translates to a range of 1194us which when applied to a pwm servo signal equals 903-2097us. At +/-100% travel, the data range is equivalent to a servo position data range of approximately 341 to 1707 which translated to pwm equals 1102 us to 1898 us.

## Contributors
D Evangelista

## Acknowledgements and references
1. The code is based on an earlier Spektrum code by M Nentwig, [https://os.mbed.com/users/offroad/code/spektRx/], however, that code appears to have used an earlier version of Spektrum standard that appears no longer to cover the serial Satellite and DSMX quad race receivers. 

2. "Spektrum Remote Receiver Interfacing", rev A, dtd 12 Apr 2016 [https://www.spektrumrc.com/ProdInfo/Files/Remote%20Receiver%20Interfacing%20Rev%20A.pdf], however, there are some slight differences between what is described there and what was observed when probing a SPM9745 DSMX remote receiver.  



CHANGES.txt
2018-Sep-15 -- v0.0.0 "Plebe"
2018-Nov-15 -- v1.0.0 "Youngster", appears to be working with SPM9745 
2018-Nov-17 -- v1.0.1 minor edits


TESTS/baud_rate_probe/main.cpp
/*
I did this to check if the LPC1768 was happy at 125000. 
It appears unhappy at 125000 and works better at 115200. 
 */

#include "mbed.h"
#include "rtos.h"

Serial pc(USBTX,USBRX,115200);

Serial receiver(p13,p14,115200);
unsigned char c; 

int main(void){
  pc.printf("hello\r\n");
  
  while(1){
    c = receiver.getc();

    if ((c==0xa2) || (c==0xb2))
      pc.printf("start\r\n");
  }
}


TESTS/receiver/simple/main.cpp
/*
  TESTS/receiver/simple/main.cpp
  Spektrum receiver simple test
  D Evangelista, 2018

  Appears to be working, 11/15/2018 with SPM9745

  NB: The specification for Spektrum Remote Receiver Interfacing,
  rev A, 12 Apr 2016, has a note about servo position ranges:
  "A full range of servo position data ranges from 0-1024 or 0-2045 depending
  on the bind type. These limits are equivalent to a +/-150% travel setting
  in AirWare. This full range translates to a range of 1194us which when 
  applied to a pwm servo signal equals 903-2097us. At +/-100% travel, the
  data range is equivalent to a servo position data range of approximately 
  341 to 1707 which translated to pwm equals 1102 us to 1898 us."
*/

#include "mbed.h"
#include "rtos.h"
#include "Spektrum.h"
#include "unity.h"

Serial pc(USBTX, USBRX, 115200);
Spektrum rx(p13, p14); 
char c;
int i;
int j; 
uint64_t now; 


int main(void){

  pc.printf("Spektrum library version ");
  pc.printf(SPEKTRUM_VERSION); 
  pc.printf("\r\nSpektrum receiver simple test\r\n");
  pc.printf("Be sure to connect orange to p13, gray to p14, blk to gnd\r\n");
  pc.printf("Receiver should have already been bound to transmitter DSMX 22ms\r\n");
  pc.printf("Printing about 22s of data now...\r\n");



  // loop to show 22s of receiver data
  for (i=0; i<1000; i++){
    now = rtos::Kernel::get_ms_count();
    
    pc.printf("%d ",i);
    if (rx.valid){
      pc.printf("%d %02x ",rx.fades,rx.system); 
      for (j=0; j<9; j++){ // my DX9 only has 9 channels
	pc.printf("%04d ",rx.channel[j]);
      }
      pc.printf("\r\n");
    }
    else
      pc.printf(" invalid\r\n");
    
    ThisThread::sleep_until(now+rx.period_ms); 
  }//for 22 s of data. 



  
  pc.printf("Successful (y/n)? ");
  pc.scanf(" %c",&c);
  TEST_ASSERT_EQUAL_MESSAGE('y',c,"receiver simple test failed.\r\n");

} // main() for TESTS/receiver/simple











TESTS/receiver/sailbot/main.cpp
/*
  TESTS/receiver/sailbot/main.cpp
  Spektrum receiver passthru test for Sailbot
  D Evangelista, 2018

  Appears to be working 11/15/2018 with SPM9745
  NB: The specification for Spektrum Remote Receiver Interfacing,
  rev A, 12 Apr 2016, has a note about servo position ranges:
  "A full range of servo position data ranges from 0-1024 or 0-2045 depending
  on the bind type. These limits are equivalent to a +/-150% travel setting
  in AirWare. This full range translates to a range of 1194us which when 
  applied to a pwm servo signal equals 903-2097us. At +/-100% travel, the
  data range is equivalent to a servo position data range of approximately 
  341 to 1707 which translated to pwm equals 1102 us to 1898 us."
*/

#include "mbed.h"
#include "rtos.h"
#include "Spektrum.h"
#include "unity.h"

Serial pc(USBTX, USBRX, 115200);
Spektrum rx(p13, p14);
PwmOut mainsail(p21);
PwmOut rudder(p22);


int main(void){
  int i=0; // a counter for for loop
  uint64_t now; // for timing the loop to be 11 or 22ms

  // setup
  pc.printf("Spektrum library version ");
  pc.printf(SPEKTRUM_VERSION);
  pc.printf("\r\nSpektrum receiver passthru test for Sailbot\r\n");
  pc.printf("Be sure to connect orange to p13, gray to p14, blk to gnd\r\n");
  pc.printf("Receiver should have already been bound to transmitter DSMX 22ms\r\n");

  pc.printf("Initializing PwmOut objects for mainsail p21 and rudder p22\r\n");
  mainsail.period_ms(20); mainsail.pulsewidth_us(1500); // amidships
  rudder.period_ms(20); rudder.pulsewidth_us(1500); // amidships

  pc.printf("Now starting while(1) loop...\r\n");
  while(1){
    now = rtos::Kernel::get_ms_count();
    
    if (rx.valid){
      // display on screen? may be too slow for 22 ms update rate
      pc.printf("%d. mainsail %4d, rudder %4d\r\n",
		i++, 
		rx.pulsewidth[0], // DX9 channel 0 throttle
		rx.pulsewidth[3]); // DX9 channel 3 rudder

      // set pwms 
      mainsail.pulsewidth_us(rx.pulsewidth[0]);
      rudder.pulsewidth_us(rx.pulsewidth[3]);
    } // if valid
    else {
      pc.printf("%d. invalid\r\n",i);
    } // not valid
    
    ThisThread::sleep_until(now+rx.period_ms); 
  }//for

  //pc.printf("Successful (y/n)? ");
  //pc.scanf(" %c",&c);
  //TEST_ASSERT_EQUAL_MESSAGE('y',c,"receiver simple test failed.\r\n");
  return 0; 
  
} // main()











Spektrum.h
/*
  Spektrum.h
  Spektrum serial receiver mbed library
  Dennis Evangelista, 2018
*/

#ifndef SPEKTRUM_H
#define SPEKTRUM_H
#define SPEKTRUM_VERSION "1.0.1"

#include "mbed.h"
#include "rtos.h"

// bind modes
#define SPEKTRUM_INT_DSMX_22MS 7
#define SPEKTRUM_EXT_DSMX_22MS 8
#define SPEKTRUM_INT_DSMX_11MS 9
#define SPEKTRUM_EXT_DSMX_11MS 10
// DSM2 bind modes not recommended, not implemented
// EXT(ernal) bind modes don't really work on solitary satellite receiver

// field definitions
// #define SPEKTRUM_MASK_1024_CHANID 0xfc00
// #define SPEKTRUM_MASK_1024_SXPOS 0x03ff
// first two only used in DSM2 which are not implemented
#define SPEKTRUM_MASK_2048_CHANID 0x7800
#define SPEKTRUM_MASK_2048_CHANID_MSB 0x78
#define SPEKTRUM_MASK_2048_SXPOS 0x07ff

// allowable system field values
// #define SPEKTRUM_22MS_1024_DSM2 0x01
// #define SPEKTRUM_11MS_2048_DSM2 0x12
#define SPEKTRUM_22MS_2048_DSMX 0xa2
#define SPEKTRUM_11MS_2048_DSMX 0xb2

#define SPEKTRUM_BAUD 115200
// Spektrum baud is supposed to be 125000, but the LPC1768 seems not
// to support nonstandard baud rates.

#define SPEKTRUM_SERVOS 7
#define SPEKTRUM_PACKET_SIZE 16
#define SPEKTRUM_CHANNELS 16
#define SPEKTRUM_COUNT2US(x) (x*600/1024+900)




/** Spektrum receiver object for connecting to eg SPM9745 receiver
 */
class Spektrum{
 public:
  /** Number of fades (failed packets) from receiver */
  unsigned int fades;
  
  /** Tells if system is in DSMX 11ms or 22ms for example */
  unsigned int system; 

  /** Contains 0-2048 values for all channels 0-15 */ 
  unsigned int channel[SPEKTRUM_CHANNELS];

  /** Contains approx 900-2100us pulsewidths corresponding to chan 0-15 */
  unsigned int pulsewidth[SPEKTRUM_CHANNELS];

  /** If true, data is value */ 
  bool valid;  // TODO switch to EventFlags?

  /** 11 or 22 ms */ 
  unsigned int period_ms;
  
  Spektrum(PinName tx, PinName rx); // constructor
  ~Spektrum(); // destructor

 private:
  UARTSerial _rx;
  unsigned char _buf[SPEKTRUM_PACKET_SIZE]; 
  Thread _packet_thread;
  void _packet_callback(void); 
};







/** For binding a Spektrum receiver to transmitter */ 
class BindPlug{
 public:
  /** e.g. SPEKTRUM_INT_DSMX_11MS, SPEKTRUM_INT_DSMX_22MS */
  int mode;
  
  BindPlug(PinName tx, PinName rx, int mode = SPEKTRUM_INT_DSMX_11MS);
  ~BindPlug();
  void bind();

 private:
  DigitalOut _3Vpin;
  DigitalOut _datapin;
};

/* LATER
class SpektrumTestDevice{
 public:
  unsigned int fades;
  unsigned int servo[7];
  SpektrumTestDevice(PinName tx, PinName rx);
  ~SpektrumTestDevice();

 private:
  Serial _receiver; 
};
*/

#endif


spektrum.h
/*
  spektrum.h
  Spektrum serial receiver mbed library
  Dennis Evangelista, 2018
*/

#ifndef SPEKTRUM_H
#define SPEKTRUM_H
#define SPEKTRUM_VERSION "1.0.0"

#include "mbed.h"
#include "rtos.h"

// bind modes
#define SPEKTRUM_INT_DSMX_22MS 7
#define SPEKTRUM_EXT_DSMX_22MS 8
#define SPEKTRUM_INT_DSMX_11MS 9
#define SPEKTRUM_EXT_DSMX_11MS 10
// DSM2 bind modes not recommended, not implemented
// EXT(ernal) bind modes don't really work on solitary satellite receiver

// field definitions
// #define SPEKTRUM_MASK_1024_CHANID 0xfc00
// #define SPEKTRUM_MASK_1024_SXPOS 0x03ff
// first two only used in DSM2 which are not implemented
#define SPEKTRUM_MASK_2048_CHANID 0x7800
#define SPEKTRUM_MASK_2048_CHANID_MSB 0x78
#define SPEKTRUM_MASK_2048_SXPOS 0x07ff

// allowable system field values
// #define SPEKTRUM_22MS_1024_DSM2 0x01
// #define SPEKTRUM_11MS_2048_DSM2 0x12
#define SPEKTRUM_22MS_2048_DSMX 0xa2
#define SPEKTRUM_11MS_2048_DSMX 0xb2

#define SPEKTRUM_BAUD 115200
// Spektrum baud is supposed to be 125000, but the LPC1768 seems not
// to support nonstandard baud rates.

#define SPEKTRUM_SERVOS 7
#define SPEKTRUM_PACKET_SIZE 16
#define SPEKTRUM_CHANNELS 16
#define SPEKTRUM_COUNT2US(x) (x*600/1024+900)




/** Spektrum receiver object for connecting to eg SPM9745 receiver
 */
class Spektrum{
 public:
  /** Number of fades (failed packets) from receiver */
  unsigned int fades;
  
  /** Tells if system is in DSMX 11ms or 22ms for example */
  unsigned int system; 

  /** Contains 0-2048 values for all channels 0-15 */ 
  unsigned int channel[SPEKTRUM_CHANNELS];

  /** Contains approx 900-2100us pulsewidths corresponding to chan 0-15 */
  unsigned int pulsewidth[SPEKTRUM_CHANNELS];

  /** If true, data is value */ 
  bool valid;  // TODO switch to EventFlags?

  /** 11 or 22 ms */ 
  unsigned int period_ms;
  
  Spektrum(PinName tx, PinName rx); // constructor
  ~Spektrum(); // destructor

 private:
  UARTSerial _rx;
  unsigned char _buf[SPEKTRUM_PACKET_SIZE]; 
  Thread _packet_thread;
  void _packet_callback(void); 
};







/** For binding a Spektrum receiver to transmitter */ 
class BindPlug{
 public:
  /** e.g. SPEKTRUM_INT_DSMX_11MS, SPEKTRUM_INT_DSMX_22MS */
  int mode;
  
  BindPlug(PinName tx, PinName rx, int mode = SPEKTRUM_INT_DSMX_11MS);
  ~BindPlug();
  void bind();

 private:
  DigitalOut _3Vpin;
  DigitalOut _datapin;
};

/* LATER
class SpektrumTestDevice{
 public:
  unsigned int fades;
  unsigned int servo[7];
  SpektrumTestDevice(PinName tx, PinName rx);
  ~SpektrumTestDevice();

 private:
  Serial _receiver; 
};
*/

#endif


Spektrum.cpp
/*
  Spektrum.cpp
  Implementation file for Spektrum serial receiver mbed library
  Dennis Evangelista, 2018
*/

#include "mbed.h"
#include "rtos.h"

#include "Spektrum.h"






/** Spektrum is used for a working connection to a Spektrum Satellite receiver
    @param tx orange wire, 3.3V supply pin (should be held high normally)
    @param rx gray wire, rx pin from the receiver
    The black wire should be connected to ground. 

    The receiver should first be bound using a BindPlug object. 
*/
Spektrum::Spektrum(PinName tx, PinName rx):
  _rx(tx, rx, SPEKTRUM_BAUD){
  
  // receiver uses e.g. p13, p14 and 115200 baud. 
  _rx.set_blocking(false); // want receiver to not block
  period_ms = 11; // start with 11ms period.

  // initialize public variables... 
  system = 0; // should be 0xa2 or 0xb2 for DSMX modes
  fades = 0;
  for (int i=0; i<SPEKTRUM_CHANNELS; i++){
    channel[i] = 0; 
    pulsewidth[i] = 0;
  }
  valid = false; 

  // start the packet reading thread
  _packet_thread.start(callback(this,&Spektrum::_packet_callback)); 
} // Spektrum(tx, rx) constructor




/** Destructor for Spektrum object 
 */
Spektrum::~Spektrum(){
} // ~Spektrum() destructor



/** Private callback for when a packet is received
 */
void Spektrum::_packet_callback(void){
  // local variables
  int count; // used to get error code -11 or num of bytes _rx.read()
  uint64_t now; // for getting 11 or 22 ms period via ThisThread::sleep_until()
  unsigned int i; // counter for for loop
  unsigned int data; // for assembling 2 bytes into uint16_t
  unsigned int chanid; // for decoding channel ID with mask 0x7800
  unsigned int servopos; // for decoding servo value with mask 0x07ff
  
  // setup
  debug("Spektrum::_packet_thread started\r\n");

  // loop
  _rx.sync(); // flush buffer
  while(1){
    now = rtos::Kernel::get_ms_count(); // for timing
    count = _rx.read(_buf,SPEKTRUM_PACKET_SIZE); // try to read packet
    
    if (count == SPEKTRUM_PACKET_SIZE){
      // got a full sized packet
      if (_buf[1] == SPEKTRUM_22MS_2048_DSMX){
	period_ms = 22; 
	valid = true;
      } // got 22ms packets
      else if (_buf[1] == SPEKTRUM_11MS_2048_DSMX){
	period_ms = 11;
	valid = true;
      } // got 11ms packets
      else
	// if system is not 0xa2 or 0xb2, treat as invalid
	valid = false;
    } // if count == 16
    else {
      // count wasn't 16 so some kind of error
      valid = false;
      _rx.sync(); // not getting enough bytes, so sync()
    }

    if (valid){
      // got a valid packet so parse it 
      fades = _buf[0];
      system = _buf[1];
      for (i=0; i<SPEKTRUM_SERVOS; i++){
	data = (_buf[2*i+2]<<8) | _buf[2*i+2+1];
	chanid = (data & SPEKTRUM_MASK_2048_CHANID) >> 11;
	servopos = data & SPEKTRUM_MASK_2048_SXPOS;
	channel[chanid] = servopos;
	pulsewidth[chanid] = SPEKTRUM_COUNT2US(servopos); 
      } // for each servo in packet
    } // if(valid)

    ThisThread::sleep_until(now+period_ms); // sleep to get right rate
  } // while(1)
} // _packet_callback() 









/** BindPlug is used to bind a Spektrum Satellite receiver
    @param tx orange wire, 3.3V supply pin, here used as a DigitalOut
    @param rx gray wire, rx pin, here used as a DigitalOut
    @param mode is mode, e.g. internal or external, DSM2 or DSMX, 11 or 22ms
    The black wire should be connected to ground.

    Default mode is internal, DSMX, 11 ms. Once created, this object will 
    send a number of falling pulses over the 3.3V supply pin to trigger 
    the satellite receiver to go into bind mode. 
*/
BindPlug::BindPlug(PinName tx, PinName rx, int mode): _3Vpin(tx),_datapin(rx){
  int i; // counter
  
  // within 200 ms of applying power, supply a bunch of falling pulses
  // according to table in Spektrum docs, most likely 9 pulses for
  // internal mode, DSMX, 11 ms.
  _3Vpin = 0;
  _datapin = 0; 
  wait(1);
  _3Vpin = 1; 
  _datapin = 1;
  wait_ms(72);
  debug("pulse ");
  for(i=0; i<mode; i++){
    debug("%d ",i);
    wait_us(116);
    _datapin = 0; // this is the falling pulse
    wait_us(116);
    _datapin = 1; 
  }
  debug("\r\n"); 
} // BindPlug(bind, mode) constructor

/** Destructor for BindPlug object
 */
BindPlug::~BindPlug(){
} // ~BindPlug() destructor






/* LATER
SpektrumTestDevice::SpektrumTestDevice(PinName tx, PinName rx): _receiver(tx,rx){
  _receiver.baud(SPEKTRUM_BAUD);
} // SpektrumTestDevice(tx, rx) constructor

SpektrumTestDevice::~SpektrumTestDevice(){
} // ~SpektrumTestDevice() destructor
*/


spektrum.cpp
/*
  spektrum.cpp
  Implementation file for Spektrum serial receiver mbed library
  Dennis Evangelista, 2018
*/

#include "mbed.h"
#include "rtos.h"

#include "spektrum.h"






/** Spektrum is used for a working connection to a Spektrum Satellite receiver
    @param tx orange wire, 3.3V supply pin (should be held high normally)
    @param rx gray wire, rx pin from the receiver
    The black wire should be connected to ground. 

    The receiver should first be bound using a BindPlug object. 
*/
Spektrum::Spektrum(PinName tx, PinName rx):
  _rx(tx, rx, SPEKTRUM_BAUD){
  
  // receiver uses e.g. p13, p14 and 115200 baud. 
  _rx.set_blocking(false); // want receiver to not block
  period_ms = 11; // start with 11ms period.

  // initialize public variables... 
  system = 0; // should be 0xa2 or 0xb2 for DSMX modes
  fades = 0;
  for (int i=0; i<SPEKTRUM_CHANNELS; i++){
    channel[i] = 0; 
    pulsewidth[i] = 0;
  }
  valid = false; 

  // start the packet reading thread
  _packet_thread.start(callback(this,&Spektrum::_packet_callback)); 
} // Spektrum(tx, rx) constructor




/** Destructor for Spektrum object 
 */
Spektrum::~Spektrum(){
} // ~Spektrum() destructor



/** Private callback for when a packet is received
 */
void Spektrum::_packet_callback(void){
  // local variables
  int count; // used to get error code -11 or num of bytes _rx.read()
  uint64_t now; // for getting 11 or 22 ms period via ThisThread::sleep_until()
  unsigned int i; // counter for for loop
  unsigned int data; // for assembling 2 bytes into uint16_t
  unsigned int chanid; // for decoding channel ID with mask 0x7800
  unsigned int servopos; // for decoding servo value with mask 0x07ff
  
  // setup
  debug("Spektrum::_packet_thread started\r\n");

  // loop
  _rx.sync(); // flush buffer
  while(1){
    now = rtos::Kernel::get_ms_count(); // for timing
    count = _rx.read(_buf,SPEKTRUM_PACKET_SIZE); // try to read packet
    
    if (count == SPEKTRUM_PACKET_SIZE){
      // got a full sized packet
      if (_buf[1] == SPEKTRUM_22MS_2048_DSMX){
	period_ms = 22; 
	valid = true;
      } // got 22ms packets
      else if (_buf[1] == SPEKTRUM_11MS_2048_DSMX){
	period_ms = 11;
	valid = true;
      } // got 11ms packets
      else
	// if system is not 0xa2 or 0xb2, treat as invalid
	valid = false;
    } // if count == 16
    else {
      // count wasn't 16 so some kind of error
      valid = false;
      _rx.sync(); // not getting enough bytes, so sync()
    }

    if (valid){
      // got a valid packet so parse it 
      fades = _buf[0];
      system = _buf[1];
      for (i=0; i<SPEKTRUM_SERVOS; i++){
	data = (_buf[2*i+2]<<8) | _buf[2*i+2+1];
	chanid = (data & SPEKTRUM_MASK_2048_CHANID) >> 11;
	servopos = data & SPEKTRUM_MASK_2048_SXPOS;
	channel[chanid] = servopos;
	pulsewidth[chanid] = SPEKTRUM_COUNT2US(servopos); 
      } // for each servo in packet
    } // if(valid)

    ThisThread::sleep_until(now+period_ms); // sleep to get right rate
  } // while(1)
} // _packet_callback() 









/** BindPlug is used to bind a Spektrum Satellite receiver
    @param tx orange wire, 3.3V supply pin, here used as a DigitalOut
    @param rx gray wire, rx pin, here used as a DigitalOut
    @param mode is mode, e.g. internal or external, DSM2 or DSMX, 11 or 22ms
    The black wire should be connected to ground.

    Default mode is internal, DSMX, 11 ms. Once created, this object will 
    send a number of falling pulses over the 3.3V supply pin to trigger 
    the satellite receiver to go into bind mode. 
*/
BindPlug::BindPlug(PinName tx, PinName rx, int mode): _3Vpin(tx),_datapin(rx){
  int i; // counter
  
  // within 200 ms of applying power, supply a bunch of falling pulses
  // according to table in Spektrum docs, most likely 9 pulses for
  // internal mode, DSMX, 11 ms.
  _3Vpin = 0;
  _datapin = 0; 
  wait(1);
  _3Vpin = 1; 
  _datapin = 1;
  wait_ms(72);
  debug("pulse ");
  for(i=0; i<mode; i++){
    debug("%d ",i);
    wait_us(116);
    _datapin = 0; // this is the falling pulse
    wait_us(116);
    _datapin = 1; 
  }
  debug("\r\n"); 
} // BindPlug(bind, mode) constructor

/** Destructor for BindPlug object
 */
BindPlug::~BindPlug(){
} // ~BindPlug() destructor






/* LATER
SpektrumTestDevice::SpektrumTestDevice(PinName tx, PinName rx): _receiver(tx,rx){
  _receiver.baud(SPEKTRUM_BAUD);
} // SpektrumTestDevice(tx, rx) constructor

SpektrumTestDevice::~SpektrumTestDevice(){
} // ~SpektrumTestDevice() destructor
*/


main.cpp
/*
  Spektrum receiver passthru test for Sailbot
  D Evangelista, 2018

  Appears to be working 11/15/2018 with SPM9745
  NB: The specification for Spektrum Remote Receiver Interfacing,
  rev A, 12 Apr 2016, has a note about servo position ranges:
  "A full range of servo position data ranges from 0-1024 or 0-2045 depending
  on the bind type. These limits are equivalent to a +/-150% travel setting
  in AirWare. This full range translates to a range of 1194us which when 
  applied to a pwm servo signal equals 903-2097us. At +/-100% travel, the
  data range is equivalent to a servo position data range of approximately 
  341 to 1707 which translated to pwm equals 1102 us to 1898 us."
*/

#include "mbed.h"
#include "rtos.h"
#include "spektrum.h"

Serial pc(USBTX, USBRX, 115200);
Spektrum rx(p13, p14);
PwmOut mainsail(p21);
PwmOut rudder(p22);


int main(void){
  int i; // a counter for for loop
  uint64_t now; // for timing the loop to be 11 or 22ms

  // setup
  pc.printf("Spektrum library version ");
  pc.printf(SPEKTRUM_VERSION);
  pc.printf("\r\nSpektrum receiver passthru test for Sailbot\r\n");
  pc.printf("Be sure to connect orange to p13, gray to p14, blk to gnd\r\n");
  pc.printf("Receiver should have already been bound to transmitter DSMX 22ms\r\n");

  pc.printf("Initializing PwmOut objects for mainsail p21 and rudder p22\r\n");
  mainsail.period_ms(20); mainsail.pulsewidth_us(1500); // amidships
  rudder.period_ms(20); rudder.pulsewidth_us(1500); // amidships

  pc.printf("Now starting while(1) loop...\r\n");
  while(1){
    now = rtos::Kernel::get_ms_count();
    
    if (rx.valid){
      // display on screen? may be too slow for 22 ms update rate
      pc.printf("%d. mainsail %4d, rudder %4d\r\n",
		i, 
		rx.pulsewidth[0], // DX9 channel 0 throttle
		rx.pulsewidth[3]); // DX9 channel 3 rudder

      // set pwms 
      mainsail.pulsewidth_us(rx.pulsewidth[0]);
      rudder.pulsewidth_us(rx.pulsewidth[3]);
    } // if valid
    else {
      pc.printf("%d. invalid\r\n",i);
    } // not valid
    
    ThisThread::sleep_until(now+rx.period_ms); 
  }//for
  
} // main()









